crates/biome_js_analyze/src/utils/restricted_glob.rs

```diff
@@ -0,0 +1,154 @@
+use biome_deserialize_macros::Deserializable;
+
+/// A restricted glov pattern only supports the following syntaxes:
+///
+/// - star `*` that matches zero or more character inside a path segment
+/// - globstar `**` that matches zero or more path segments
+/// - Use `\*` to escape `*`
+/// - `?`, `[`, `]`, `{`, and `}` must be escaped using `\`.
+///   These characters are reserved for future use.
+/// - `!` must be escaped if it is the first characrter of the pattern
+///
+/// A path segment is delimited by path separator `/` or the start/end of the path.
+#[derive(Clone, Debug, Deserializable, serde::Deserialize, serde::Serialize)]
+#[serde(try_from = "String", into = "String")]
+pub struct RestrictedGlob(globset::GlobMatcher);
+
+impl std::ops::Deref for RestrictedGlob {
+    type Target = globset::GlobMatcher;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::fmt::Display for RestrictedGlob {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let repr = self.0.glob().to_string();
+        f.write_str(&repr)
+    }
+}
+
+impl From<RestrictedGlob> for String {
+    fn from(value: RestrictedGlob) -> Self {
+        value.to_string()
+    }
+}
+
+impl std::str::FromStr for RestrictedGlob {
+    type Err = globset::ErrorKind;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        is_restricted_glob(value)?;
+        let mut glob_builder = globset::GlobBuilder::new(value);
+        // Allow escaping with `\` on all platforms.
+        glob_builder.backslash_escape(true);
+        // Only `**` can match `/`
+        glob_builder.literal_separator(true);
+        match glob_builder.build() {
+            Ok(glob) => Ok(RestrictedGlob(glob.compile_matcher())),
+            Err(error) => Err(error.kind().clone()),
+        }
+    }
+}
+
+impl TryFrom<String> for RestrictedGlob {
+    type Error = globset::ErrorKind;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        value.parse()
+    }
+}
+
+#[cfg(feature = "schemars")]
+impl schemars::JsonSchema for RestrictedGlob {
+    fn schema_name() -> String {
+        "Regex".to_string()
+    }
+
+    fn json_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+        String::json_schema(gen)
+    }
+}
+
+/// Returns an error if `pattern` doesn't follow the restricted glob syntax.
+fn is_restricted_glob(pattern: &str) -> Result<(), globset::ErrorKind> {
```

Comment:
When handling an error coming from a third party library, we should use our diagnostics instead. You should create an adapter, like in this case:

https://github.com/biomejs/biome/blob/5579f9fe74ab6955e0a9e81155019c18e2310cea/crates/biome_diagnostics/src/adapters.rs#L114-L136

And you can create it like in this case:

https://github.com/biomejs/biome/blob/5579f9fe74ab6955e0a9e81155019c18e2310cea/crates/biome_configuration/src/diagnostics.rs#L116-L124

Reply:
I improved the code by completely encapsulating `gloset`. This will ease potential migrations. I also created a dedicated error type for RestrictedRegex.

The newly introduced error types keep track of the location of the error. I used a custom implementation of `Deserializable` to report the error location.

I don't see the advantage of implementing `biome_diagnostic::Diagnostic` because the error is reported upon deserilization and must be compatible with Biome and Serde. Correct me if I am wrong here.

Reply:
The point of the adapters is to make errors coming from third party libraries compatible with our diagnostics. Third-party diagnostics can't implement the trait `Diagnostic`. The second point is that adapters is to implement the `#[source]` directive gracefully. 

The current implementation creates a lot of indirection, and it eventually:
- does a simple `format_args!`, which is what you would do inside an adapter
- you lose information because we don't know anymore where the error is coming from
___

Now, what I don't understand is why you use a third-party library **to create a new error**. The description doesn't explain that. However, if that's because you envision the use of `globset` functions that will return a `globset::ErrorKind`, then you have more reasons to create an adapter.

**If that's the case**, then we should add a new optional `source` to `DeserializationDiagnostic`, and create a function called `with_source`, so we store the `globset` error.

**If it's _not_ the case**, there's no need of using `globset::ErrorKind` and just a string is sufficient.

Reply:
> you lose information because we don't know anymore where the error is coming from

I don't get it.
Do you mean the location in the code?
Because we have a precise location of the error in the config file.

> does a simple format_args!, which is what you would do inside an adapter

What is the difference between doing it outside and doing it inside?
One of my concern is o f depending on `biome_diagnostic` where we don't need to.
This reduces code portability.
I don't understand the value of `biome_diagnostic` for simple errors.
I think I don't have a clear view of the benefit of `biome_diagnostic`.

> Now, what I don't understand is why you use a third-party library to create a new error.

I did this for convenience: it avoids writing more code.
I think now that we have our own error type it might be better to turn `globset::ErrorKind` into `RestrictedGlobError`.